configurator/data/run.sh

```diff
@@ -6,6 +6,21 @@ ENFORCE_BASEPATH=$(bashio::config 'enforce_basepath')
 IGNORE_PATTERN="$(bashio::jq "/data/options.json" ".ignore_pattern")"
 WAIT_PIDS=()
 
+# If an SSH key is defined in the configuration options, add it for use by git
+if bashio::config.has_value "ssh_key"; then
+  bashio::log.info "Adding SSH key, for use by git"
+
+  # Start the SSH agent and add the key
+  eval $(ssh-agent -s)
+  KEY=$(bashio::config 'ssh_key')
+  ssh-add -q <(echo "$KEY")
```

Comment:
According to the schema in `config.json` this a list of keys?

Reply:
This is again a straight copy from the git_pull add-on. The lines of a single key are split up into a JSON array. See the example here: https://www.home-assistant.io/addons/git_pull/ . Multi-line strings are not supported in JSON, so it looks like we're stuck with either an array, or a very long string on a single line.

I've just tried passing a key to `ssh-add` with all the newlines removed (both with & without the header & footer lines), and it said the format was invalid. I think this means that if you wanted to put the whole key in a single JSON string you would have to add "\n" control characters. I think the method used by git_pull might be the best we can do.

In any case, I think it makes sense for the format of the option to match the format used by git_pull, since the two add-ons are likely to be used together if the configuration is managed in git.

Reply:
You need address this before we can merge it. If they is an array, you need loop over it

Reply:
I'm sorry, but I'm really not sure what you'd like me to do here - it's not an array of keys, it's a single key split into an array of lines. The approach is exactly the same as used in the git_pull add-on in this same repository.

I've tested the script and it successfully adds the key to the agent. Is something more needed? I'd be happy to make changes, but I'd need more details.

Reply:
@adamgreg You'll need to loop over the keys:

Example: <https://github.com/home-assistant/hassio-addons/blob/6af0f83c55bb9db3bdff9524cf3f0ec361148baf/ssh/data/run.sh#L11>

Reply:
It's only 1 key, and the code works. Nothing to loop over.

The ssh add-on deals with a list of (short) authorised key strings. The code in this PR (as well as the code in this repo's git_pull add-on, which I assume is acceptable) deals with a single (long) private key, split at each newline into an array of strings. Honestly, the script works.

Reply:
Code that works does not have to be implemented correctly @adamgreg. This has to be adjusted as pointed out now a couple of time.

Let me lay it out for you:

You have specified in the add-on config, that a list of keys can be provided:

https://github.com/home-assistant/hassio-addons/blob/5796b073284fa477f8bcf88b3dbb90918d3f34c2/configurator/config.json#L45

However, in the code, only 1 item is picked and the code does not take into account the configuration can contain a list of items. A list needs to be iterated.

Does it work if you enter a single key? Sure. Is the list handled? Nope.

I'm currently unsure what the problem is at this point. You only have to modify 3 lines, as I already kindly provided by linking the SSH add-on example.

Reply:
The schema specifies a list of strings, not a list of keys. It's a single key split by line into an array. It does not support multiple keys. It just works the same way as the git_pull plugin.

Reply:
that is correct

Reply:
Thanks! I gave a bit more of an explanation in my first comment: https://github.com/home-assistant/hassio-addons/pull/592#discussion_r289639143 . The private keys are just a bit too long to paste into a single JSON string!

Bashio/jq does a great job here of getting the whole key from JSON, with whitespace between the lines - just right for ssh-add.

Having a key split like that in the config doesn't look great, but I think it's better than squeezing it onto one line. It's good to be consistent with the format of the config of the git_pull add-on, so at least people who use both (like my client) can copy & paste the key between them.

Reply:
Here's a blog post I just found about this approach to storing keys and other long strings in JSON, that might be helpful: https://www.gun.io/blog/multi-line-strings-in-json

Reply:
I get that, although, JSON can still do that using newline char `\n`.
So how would I add multiple keys in that case?

Reply:
Second thought, why not convert this in to a simple file list?
e.g., point directly to a key `/ssl/my.key` or `/config/ssh/id_rsa`.

Reply:
Ok, that does sound like a good idea actually. I've made the change.